### ISSUE TITLE
masters causing route53 rate exceeded/Throttling issue in AWS

### ISSUE THREAD

**Issue by hridyeshpant** [kubernetes/kops#1376]
Kubernetes version (use kubectl version):

kops version
Version 1.4.4

kubectl version
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:22:15Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}

Environment:

Cloud provider or hardware configuration: aws
OS (e.g. from /etc/os-release): amazon linux
What happened:
as per
http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html

Five requests per second per AWS account are allowed. If you submit more than five requests per second, Amazon Route 53 returns an HTTP 400 error (Bad request). The response header also includes a Code element with a value of Throttling and a Message element with a value of Rate exceeded.
we set up kubernetes in AWS with having three master using kops http://kubernetes.io/docs/getting-started-guides/kops/.
All three master EC2 instances which makes the R53 API calls which is exceeding the 5API calls per second. Here instance id and API calls per second.

i-033a2d4dff5fa39d3 ** Max 3 API Calls/sec
i-008dcee283bbc7c5a ** Max 2.9 API Calls/sec
i-099dbb226a9fe25e6 ** Max 2.9 API Calls/sec
All this sources making the API call "ListResourceRecordSets".
Due to this we are continuously getting Rate exceeded and causing our other deployment failed during route53 operation .After deleting the masters things get normal.
What you expected to happen:
Can you suggest what is the best possible solution for this? and how we can prevent this.

How to reproduce it (as minimally and precisely as possible):
create multiple master with kops.

as per https://github.com/kubernetes/kubernetes/issues/39509#issuecomment-271004716, it looks like issue with kops.

**Comment by chrislovecnm** [comment]
@justinsb any ideas?

**Comment by justinsb** [comment]
So I'm trying to reproduce this, and can't with the latest build.  We'll be shipping 1.5.0-alpha1 real-soon-now, so you may just want to try that when it ships.  But I don't know of any fixes since 1.4.4 that would cause this (I'll try 1.4.4 next).

But if you want to diagnose, that would be very helpful.  There are two components that could be making the calls - protokube & dns-controller.  Both ship with pretty verbose logging by default.

If you SSH to a master, can you try this to see the protokube logs:

```
> sudo docker ps | grep protokube
673df9049490        protokube:1.5.0-alpha1                                    "/usr/bin/protokube -"   25 minutes ago      Up 25 minutes                           admiring_noether
> sudo docker logs 673df9049490
```

`sudo docker logs 413f8d461406 2>&1 | grep aws_dns | less`  shows the API calls, but it depends how old a version of kops/protokube you are running.  I did note that this makes 2 requests per minute, where it should be making one (I opened #1390).  The full `sudo docker logs` will show everything.


And this to see the dns-controller logs (you can actually do this from anywhere with kubectl access):

```
> kubectl get pods -n kube-system | grep dns-con
dns-controller-1449167825-kiy00                          1/1       Running   0          15m
>  kubectl logs -n kube-system dns-controller-1449167825-kiy00
```

I opened #1391 because the output is a little hard to decipher, but you're looking for `Updating resource record`

**Comment by justinsb** [comment]
I also went back to 1.4.4 and created a cluster with `~/bin/kops-1.4.4 create cluster --zones us-east-1b,us-east-1c,us-east-1d --master-zones us-east-1b,us-east-1c,us-east-1d k144.example.com`

I saw the same behaviour - about 2 requests / minute / master, which is too much, but still nowhere near the rate you are reporting (unless those rates are actually in minutes, not seconds).

Also, can you provide the log message showing the route53 rate limiting error?

**Comment by hridyeshpant** [comment]
@justinsb i am attaching logs for both "dns-controller" and "protokube".
i can see protokube logs making too many calls to route53 per second.let me know if you need more info. For now i am deleting the cluster as i am start getting rate exceeded issue. 
[kuberlog.zip](https://github.com/kubernetes/kops/files/692591/kuberlog.zip)

**Comment by hridyeshpant** [comment]
Even i can see dns-controller log showing Throttling: Rate exceeded error some time.

I0109 01:17:39.708087       1 dnscontroller.go:460] Update node/ip-172-20-101-9.us-west-2.compute.internal: [{A node/ip-172-20-101-9.us-west-2.compute.internal/internal 172.20.101.9 true} {A node/ip-172-20-101-9.us-west-2.compute.internal/external 54.186.208.183 true}]
W0109 01:17:41.283664       1 dnscontroller.go:103] Unexpected error in DNS controller, will retry: error querying for zones: Throttling: Rate exceeded
	status code: 400, request id: 6a946b0c-d609-11e6-9a90-05afbbd6eb0b

**Comment by chrislovecnm** [comment]
@hridyeshpant we have a nasty issue with rate limiting and PV that will be fixed in 1.4.8, and 1.5.2.  What do your controller logs look like?

**Comment by hridyeshpant** [comment]
@chrislovecnm please see my log file (kuberlog.zip) attached in previous post .

**Comment by chrislovecnm** [comment]
@hridyeshpant I am looking to see if you have errors in /var/log/kubernetes/kube-controller-manager.log ... I have an open PR that will be dropped in on Tuesday to address a really nasty issue with PVs.  How many PV do you have on the cluster?

**Comment by hridyeshpant** [comment]
@chrislovecnm sorry i just deleted the cluster as other deployment get blocked because of rate exceeding error.
you mean Persistent Volume  in cluster?
can you let me know how to check PV, i can create cluster again.

**Comment by chrislovecnm** [comment]
you have to create them. Did you install anything on the cluster?

**Comment by hridyeshpant** [comment]
no i haven't run anythings extra. i just ran below command 
>>
kops create cluster --v=3 --state=s3://ewe-kubernetes-test --cloud=aws     --name=kubetnets.us-west-2.test.mycomp.com     --ssh-public-key=~/.ssh/ewetest.pub     --master-size=t2.medium     --master-zones=us-west-2c,us-west-2a,us-west-2b     --network-cidr=172.20.0.0/16     --vpc=vpc-90d677f7     --node-count=3     --node-size=t2.micro     --zones=us-west-2c,us-west-2a,us-west-2b
>>

**Comment by justinsb** [comment]
I'm reasonably sure this will be fixed in 1.5.0-alpha, which I'm planning to try to push first thing tomorrow.

**Comment by hridyeshpant** [comment]
@justinsb thanks . 
Could you please update here when this fix available and how to get 1.5.0-alpha?

**Comment by justinsb** [comment]
Just pushed 1.5.0-alpha2: https://github.com/kubernetes/kops/releases/tag/1.5.0-alpha2 :-)  Finally!


If you try it, please specify `--channel alpha` when creating a cluster (`kops create cluster --channel alpha`) or run `kops upgrade cluster --channel alpha`, to get the new image (although it should just be a little slower if you don't use the newer image)

**Comment by hridyeshpant** [comment]
kops version
Version 1.5.0-alpha2

kops update cluster --channel alpha
Error: unknown flag: --channel

@justinsb

**Comment by hridyeshpant** [comment]
still getting same issue as soon as masters are up.
attaching the logs 
[kuberlog.zip](https://github.com/kubernetes/kops/files/695389/kuberlog.zip)

@justinsb

**Comment by justinsb** [comment]
I don't see us making a ridiculous number of calls any more, which is good.

A few ideas:

1) I can see that we could cache the list of HostedZones.  I'll think about this one.
2) Do you have a huge number of HostedZones in the AWS account?
3) Could you have something else also making a large number of AWS calls.  We are listing zones every ~15 seconds, which should be well under the limit (but hence 1 and 2!).  If you set up CloudTrail (free) you can see a log of all your requests, and hopefully see what all these calls are.

**Comment by hridyeshpant** [comment]
we have 10346 record set in that hosted zone. I dont think any other call making this cause. This issue start as soon as master created and disappear when i delete the cluster

also i got real issue when deleting the cluster using the 1.5.0-alpha2.
This version delete all my NAT instances, at least i can confirm old version was not deleting any such things. 

autoscaling-group:master-us-west-2c.masters.kubetnets.us-west-2.test.mycomp.com	ok
nat-gateway:nat-04f6e95d3e75265d8	ok
nat-gateway:nat-06b6add1aaf5a738d	ok
nat-gateway:nat-0d1cc880def78c447	ok
nat-gateway:nat-06de87090d3ec73d3	ok
nat-gateway:nat-0bd70454040cca579	ok
nat-gateway:nat-0591234fb69d0fb06	ok
nat-gateway:nat-09848f1451db8f1c0	ok
0110 16:25:14.157268   82226 delete_cluster.go:1872] Deleting IAM role policy "nodes.kubetnets.us-west-2.test.mycomp.com" "nodes.kubetnets.us-west-2.test.mycomp.com"
I0110 16:25:14.157413   82226 request_logger.go:45] AWS request: iam/DeleteRolePolicy
instance:i-094432b26ac96b6ce	ok
nat-gateway:nat-0decf602a5f364086	ok
instance:i-06bcf57a907bdf21c	ok

even through I was using my own VPC, why new version is deleting the nats gateway?
@justinsb

**Comment by justinsb** [comment]
Opened #1428 to track the nat gateway problem

**Comment by justinsb** [comment]
Opened https://github.com/kubernetes/kubernetes/issues/39674 for what I think is the issue there

**Comment by justinsb** [comment]
An update: I think I have found the solution - we need to support UPSERT in the DNS provider.  PR is here https://github.com/kubernetes/kubernetes/pull/39957 .  The problem is that if you have 10k resourcerecordsets we can't realistically list them without hitting rate limits because of pagination (100 records per page), and we can't query them per name because then we'll hit rate limits if we're updating multiple names at once (and we have to update multiple names at once, otherwise we hit rate limits on concurrent updates)

There are also a few minor tweaks ... we can cache the list of hosted zones, which will save some queries also.  But the big win for your case depends on 39957.

There's also still a problem in that DNS precreation may be slow in kops itself, but that is just an optimization and optional, and we have a feature flag that you can disable (`export KOPS_FEATURE_FLAG=-DNSPreCreate`.  The fix for that one is probably https://github.com/kubernetes/kubernetes/pull/39805

**Comment by hridyeshpant** [comment]
@justinsb please let me know when this fix is available to test along with NATs deletion fix.

**Comment by justinsb** [comment]
We're blocked on upstream I think, I'm punting to 1.5.1 in preparation for 1.5.0 release, but it remains a P0 there.

**Comment by chrislovecnm** [comment]
Has this been resolved?  I think it has upstream, and 1.6.x does not have these issues anymore.

**Comment by geojaz** [comment]
Definitely fixed in 1.6+ and I think also latest 1.5. closing
